src/DotNetWorker.Core/Converters/Converter/ConversionResult.cs

```diff
@@ -0,0 +1,83 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.Azure.Functions.Worker.Converters
+{
+    /// <summary>
+    /// A type representing the result of function input conversion operation.
+    /// </summary>
+    public readonly struct ConversionResult
+    {
+        // A static ConversionResult instance which represents an unhandled conversion.
+        private static readonly ConversionResult _unhandledConversionResult = new(isHandled: false);
+        
+        /// <summary>
+        /// Gets a value indicating whether the converter acted on the input to execute a conversion operation.
+        /// </summary>
+        public bool IsHandled { get; }
+
+        /// <summary>
+        /// Gets a value indicating whether the conversion operation was successful or not.
+        /// </summary>
+        public bool? IsSuccessful { get; }
```

Comment:
Yes. Those 3 are the valid states. 

Currently we expose the 3 static methods ( `Unhandled`, `Success` and `Failure`) on this type to create those 3 variations. The fourth option, one can create is using the default constructor (`new ConversionResult()`), But the properties of this instance cannot be set as they can be only set using the above mentioned methods. So essentially that newed up instance will be same as the result of `ConversionResult.Unhandled()`

Another option to reduce the confusion is to move the `IsHandled` to `IInputConverter`. So the caller first needs to explicitly check the converter can handle before calling `ConvertAsync`. In this way the `ConversionResult` only represents the real result, not a capability of the converter. feels more clean.

    if (converter.CanHandle(converterContext))
    {
        var conversionResultTask = converter.ConvertAsync(converterContext);
        ....
    }

Reply:
I like the idea of checking if the converter can handle before calling convert, that seems like a cleaner way to use this. However I don't think that's the same as what `IsHandled` is about? My understanding is that IsHandled is used to check if the converter has handled the conversion, at this stage we already know if we "can handle" the request and are past that.

As for the states, I don't think we should ever be in a state where IsHandled is false and IsSuccessful is true - essentially the 3 states that Brett pointed out above are the only ones that should be valid. Do you remember why we added IsHandled into the mix? I remember we touched on this during one of our design meetings. Was it related to tracking which conversions are queued to be handled but are not yet? I'd like to better understand the scenario where we need to know if a request was handled or not

Reply:
Alternate proposal.

```
public interface IInputConverter
{
    /// <summary>
    /// Gets a value indicating whether this converter can handle the conversion request.
    /// </summary>
    /// <returns></returns>
    bool CanConvert(ConverterContext context);

    /// <summary>
    /// Executes an input conversion operation using the converter context provided.
    /// </summary>
    ValueTask<ConversionResult> ConvertAsync(ConverterContext context);
}
```

and `IsHandled` prop will be removed from [ConversionResult](https://github.com/Azure/azure-functions-dotnet-worker/blob/9f2e6dda5ff177e69f80d471750f5ab2d0f7f6fc/src/DotNetWorker.Core/Converters/Converter/ConversionResult.cs#L9) and it will be a type reprenseting only the result of the conversion operation.

The infrastructure code which is calling the ConvertAsync first checks whether this is the right converter to use.
```
if (converter.CanConvert(converterContext))
{
    var conversionResultTask = converter.ConvertAsync(converterContext);
    ....
}
```